File: client/ios/RelatedDigitalNotificationService/NotificationService.m
//
//  NotificationService.m
//  NotificationService
//
//  Created by Egemen Gülkılık on 1.05.2024.
//

#import "NotificationService.h"
#import "FirebaseMessaging.h"
#import "RelatedDigitalNotificationService.h"

@interface NotificationService ()

@property (nonatomic, strong) void (^contentHandler)(UNNotificationContent *contentToDeliver);
@property (nonatomic, strong) UNMutableNotificationContent *bestAttemptContent;

@end

@implementation NotificationService

- (void)didReceiveNotificationRequest:(UNNotificationRequest *)request withContentHandler:(void (^)(UNNotificationContent * _Nonnull))contentHandler {
    self.contentHandler = contentHandler;
    self.bestAttemptContent = [request.content mutableCopy];
  
    if (self.bestAttemptContent) {
          NSMutableDictionary *userInfo = self.bestAttemptContent.userInfo.mutableCopy;
          NSString *emPushSp = userInfo[@"emPushSp"];
          if (emPushSp) {
              NSLog(@"emPushSp: %@", emPushSp);
              [userInfo removeObjectForKey:@"fcm_options"];
              self.bestAttemptContent.userInfo = userInfo;

              [RelatedDigitalNotificationService didReceiveNotificationRequest:@"rniostestapptest" withBestAttemptContent:self.bestAttemptContent withContentHandler:self.contentHandler];
          }
          else {
            [[FIRMessaging extensionHelper] populateNotificationContent:self.bestAttemptContent withContentHandler:contentHandler];
          }
      }
    
    // Modify the notification content here...
    //self.bestAttemptContent.title = [NSString stringWithFormat:@"%@ [modified]", self.bestAttemptContent.title];
    //self.contentHandler(self.bestAttemptContent);
}

- (void)serviceExtensionTimeWillExpire {
    // Called just before the extension will be terminated by the system.
    // Use this as an opportunity to deliver your "best attempt" at modified content, otherwise the original push payload will be used.
    self.contentHandler(self.bestAttemptContent);
}

@end

File: example/ios/RelatedDigitalExampleNotificationServiceExtension/NotificationService.swift
//
//  NotificationService.swift
//  RelatedDigitalExampleNotificationServiceExtension
//
//  Created by Egemen Gülkılık on 13.04.2023.
//

import RelatedDigitalIOS
import UserNotifications

class NotificationService: UNNotificationServiceExtension {

  var contentHandler: ((UNNotificationContent) -> Void)?
  var bestAttemptContent: UNMutableNotificationContent?

  override func didReceive(
    _ request: UNNotificationRequest,
    withContentHandler contentHandler: @escaping (UNNotificationContent) -> Void
  ) {
    self.contentHandler = contentHandler
    bestAttemptContent = (request.content.mutableCopy() as? UNMutableNotificationContent)
    DispatchQueue.main.async {
      RelatedDigital.initialize(
        organizationId: "676D325830564761676D453D", profileId: "356467332F6533766975593D",
        dataSource: "visistore", launchOptions: nil)

      RelatedDigital.enablePushNotifications(
        appAlias: "RDIOSExample", launchOptions: nil,
        appGroupsKey: "group.com.relateddigital.RelatedDigitalExample.relateddigital")
      RDPush.didReceive(self.bestAttemptContent, withContentHandler: contentHandler)
    }

  }

  override func serviceExtensionTimeWillExpire() {
    // Called just before the extension will be terminated by the system.
    // Use this as an opportunity to deliver your "best attempt" at modified content, otherwise the original push payload will be used.
    if let contentHandler = contentHandler, let bestAttemptContent = bestAttemptContent {
      contentHandler(bestAttemptContent)
    }
  }

}

